validator_client/src/duties_service.rs

```diff
@@ -674,84 +682,74 @@ async fn poll_beacon_attesters_for_epoch<T: SlotClock + 'static, E: EthSpec>(
+        .iter()
+        .filter(|pubkey| !attesters_in_response.contains(pubkey))
+        .filter_map(|pubkey| duties_service.validator_store.validator_index(pubkey))
+        .collect::<Vec<_>>();
+
+    let new_duties = if !indices_to_request.is_empty() {
+        post_validator_duties_attester(duties_service, epoch, indices_to_request.as_slice())
+            .await?
             .data
             .into_iter()
-            .filter(|duty| {
-                if duties_service.per_validator_metrics() {
-                    let validator_index = duty.validator_index;
-                    let duty_slot = duty.slot;
-                    if let Some(existing_slot_gauge) =
-                        get_int_gauge(&ATTESTATION_DUTY, &[&validator_index.to_string()])
-                    {
-                        let existing_slot = Slot::new(existing_slot_gauge.get() as u64);
-                        let existing_epoch = existing_slot.epoch(E::slots_per_epoch());
-
-                        // First condition ensures that we switch to the next epoch duty slot
-                        // once the current epoch duty slot passes.
-                        // Second condition is to ensure that next epoch duties don't override
-                        // current epoch duties.
-                        if existing_slot < current_slot
-                            || (duty_slot.epoch(E::slots_per_epoch()) <= existing_epoch
-                                && duty_slot > current_slot
-                                && duty_slot != existing_slot)
-                        {
-                            existing_slot_gauge.set(duty_slot.as_u64() as i64);
-                        }
-                    } else {
-                        set_int_gauge(
-                            &ATTESTATION_DUTY,
-                            &[&validator_index.to_string()],
-                            duty_slot.as_u64() as i64,
-                        );
-                    }
-                }
-
-                local_pubkeys.contains(&duty.pubkey) && {
-                    // Only update the duties if either is true:
-                    //
-                    // - There were no known duties for this epoch.
-                    // - The dependent root has changed, signalling a re-org.
-                    attesters.get(&duty.pubkey).map_or(true, |duties| {
-                        duties
-                            .get(&epoch)
-                            .map_or(true, |(prior, _)| *prior != dependent_root)
-                    })
-                }
-            })
+            .chain(response.data)
             .collect::<Vec<_>>()
+    } else {
+        response.data
     };
 
+    drop(fetch_timer);
+
+    let _store_timer = metrics::start_timer_vec(
+        &metrics::DUTIES_SERVICE_TIMES,
+        &[metrics::UPDATE_ATTESTERS_STORE],
+    );
+
+    if duties_service.per_validator_metrics() {
+        update_per_validator_duty_metrics::<E>(current_slot, &new_duties);
```

Comment:
Previously we were running this function for *all* duties and now we're only running it for *new* duties. I think that might pose a problem where we don't update the metrics to show the next epoch duties, once that next epoch arrives (assuming they haven't changed since the previous epoch).

I *think* we could solve this by iterating over `duties_service.attesters` instead.

Reply:
Thanks, didn't think about this scenario! Will push a fix.